README.md

This repo contains machine learning algorithms coded from scratch. These implementations are not designed for
high-performance, the focus lies on coding a generic implementation only with necessary code.


There are two blog posts about the code and the machine learning algorithms as whole.
You can find the gradient boosting one [here](https://www.statworx.com/blog/coding-gradient-boosted-machines-in-100-lines-of-code) and
the one about regression trees [here](https://www.statworx.com/blog/coding-regression-trees-in-150-lines-of-code).

This work was originally by Andre Bleier. I thought that it was a good idea but found that it had a bug that meant that
the predictions were wrong. However, this was easily fixed. I hope to made some additions to the code in the future.


NEWS.md
Saturday, November 28, 2020
rf100 Version: 0.0.1

1) put the code in R package format at parsifal9/rf100
2) added code so fitted values returned for correct obervation
   a) removed the change in reg_tree_imp -- now all of the cahnges are in sprout_tree


######################################################################################################
Saturday, November 14, 2020

the original code was bagging the data 
## - randomly sample the data with replacement (duplicate are possible)
#   train <-
#     data[sample(1:nrow(data), size = nrow(data), replace = TRUE),]
and returning the fitted values for the bagged data.
So if the data was of size $n \timesp$ then $n$ fitted values are retuned but not
for the data, they are for a bagged version of the data.

This is returned from sprout_tree and mean of the fitted values is calculated.

##calculate the final fit as a mean of all regression trees
# rf_fit <- apply(fits, MARGIN = 1, mean, na.rm = TRUE)

Now how this is wrong. We are taking the mean of the fits for random data points.

To fix this

1) in sprout_tree, record which data points are in the bootstrap sample

#temp1<-sample(1:nrow(data), size = nrow(data), replace = TRUE)
#train <-      data[temp1,]


2) send the size of the data to reg_tree_imp
#tree <- reg_tree_imp(formula = formula_new,
#                         data = train,
#                         minsize = ceiling(nrow(train) * 0.1),data.n=dim(data)[[1]])

so reg_tree_imp can return a tree$fit of the correct size.



3)  in sprout_tree, group by temp1 and get the mean y value for each leaf. This is the fitted value
#  aa<-data.frame(tree$fit,temp1) %>%
#        dplyr::group_by(temp1)%>%
#        dplyr::summarize(Mean = mean(tree.fit, na.rm=TRUE))
#    aa<-as.matrix(aa)
#    tt<-rep(NA,length(temp1))
#    tt[aa[,1]]<-aa[,2]
#    # save the fit and the importance
#    return(list(tt, tree$importance))

So it works now.

We still have the problems
* all the warnings -- `summarise()` ungrouping output (override with `.groups` argument)
* we are getting the fitted values for the samples that were selected by the bootstrap ( the in-bag samples).
Would be better if we got the fitted values for the out-of-bag samples. (is this true?)
* the "randomly sample features" is done at the same level as the "bag the data". I think it should be done at each node

R/reg_tree.R
# Regression Tree ---------------------------------------------------------

# This is the splitting criterion we minimize (SSE [Sum Of Squared Errors]):
# $SSE = \sum_{i \in S_1} (y_i - \bar(y)1)^2 + \sum_{i \in S_2} (y_i - \bar(y)2)^2$
sse_var <- function(x, y) {
  splits <- sort(unique(x))
  sse <- c()
  for (i in seq_along(splits)) {
    sp <- splits[i]
    sse[i] <- sum((y[x < sp] - mean(y[x < sp]))^2) + sum((y[x >= sp] - mean(y[x >= sp]))^2) 
  }
  split_at <- splits[which.min(sse)]
  return(c(sse = min(sse), split = split_at))
}


#' reg_tree
#' Fits a simple regression tree with SSE splitting criterion. The estimator function
#' is the mean.
#' 
#' @param formula an object of class formula
#' @param data a data.frame or matrix
#' @param minsize a numeric value indicating the minimum size of observations
#'                in a leaf
#'
#' @return \itemize{
#' \item tree - the tree object containing all splitting rules and observations
#' \item fit - our fitted values, i.e. X %*% theta
#' \item formula - the underlying formula
#' \item data - the underlying data
#' }
#' @export
#'
#' @examples # Complete runthrough see: www.github.com/andrebleier/cheapml
#'# Regression Tree Runthrough ----------------------------------------------
#'
#'# load data simulation tool
#'# library(devtools)
#'# devtools::install_github('andrebleier/Xy')
#'library(Xy)
#'
#'# simulate data
#'sim <- Xy(n = 2500, # 2500 observations
#'          numvars = c(2, 0), # 2 linear variables / 0 nonlinear
#'          noisevars = 0, # no noise variables
#'          catvars = 0, # omit dummy variables
#'          stn = 5,
#'          intercept = TRUE
#')
#'
#'# simulation overview
#'sim
#'
#'# get the formula
#'eq <- sim$eq
#'
#'# get the formula
#'model_df <- sim$data
#'
#'
#'# fit
#'mod <- reg_tree(formula = eq, data = model_df, minsize = 400)
#'
#'# our tree
#'mod$tree
#'
#'# our fitted values
#'mod$fit
#'

reg_tree <- function(formula, data, minsize) {
  
  # coerce to data.frame
  data <- as.data.frame(data)
  
  # handle formula
  formula <- terms.formula(formula)
  
  # get the design matrix
  X <- model.matrix(formula, data)
  
  # extract target
  y <- data[, as.character(formula)[2]]
  
  # initialize while loop
  do_splits <- TRUE
  
  # create output data.frame with splitting rules and observations
  tree_info <- data.frame(NODE = 1, NOBS = nrow(data), FILTER = NA, TERMINAL = "SPLIT",
                          stringsAsFactors = FALSE)
  
  # keep splitting until there are only leafs left
  while(do_splits) {
    
    # which parents have to be splitted
    to_calculate <- which(tree_info$TERMINAL == "SPLIT")
    
    for (j in to_calculate) {
      
      # handle root node
      if (!is.na(tree_info[j, "FILTER"])) {
        # subset data according to the filter
        this_data <- subset(data, eval(parse(text = tree_info[j, "FILTER"])))
        # get the design matrix
        X <- model.matrix(formula, this_data)
      } else {
        this_data <- data
      }
      
      # estimate splitting criteria
      splitting <- apply(X,  MARGIN = 2, FUN = sse_var, y = this_data[, all.vars(formula)[1]])
      
      # get the min SSE
      tmp_splitter <- which.min(splitting[1,])
      
      # define maxnode
      mn <- max(tree_info$NODE)
      
      # paste filter rules
      tmp_filter <- c(paste(names(tmp_splitter), ">=", 
                            splitting[2,tmp_splitter]),
                      paste(names(tmp_splitter), "<", 
                            splitting[2,tmp_splitter]))
      
      # Error handling! check if the splitting rule has already been invoked
      split_here  <- !sapply(tmp_filter,
                             FUN = function(x,y) any(grepl(x, x = y)),
                             y = tree_info$FILTER)
      
      # append the splitting rules
      if (!is.na(tree_info[j, "FILTER"])) {
        tmp_filter  <- paste(tree_info[j, "FILTER"], 
                             tmp_filter, sep = " & ")
      } 
      
      # get the number of observations in current node
      tmp_nobs <- sapply(tmp_filter,
                         FUN = function(i, x) {
                           nrow(subset(x = x, subset = eval(parse(text = i))))
                         },
                         x = this_data)  
      
      # insufficient minsize for split
      if (any(tmp_nobs <= minsize)) {
        split_here <- rep(FALSE, 2)
      }
      
      # create children data frame
      children <- data.frame(NODE = c(mn+1, mn+2),
                             NOBS = tmp_nobs,
                             FILTER = tmp_filter,
                             TERMINAL = rep("SPLIT", 2),
                             row.names = NULL)[split_here,]
      
      # overwrite state of current node
      tree_info[j, "TERMINAL"] <- ifelse(all(!split_here), "LEAF", "PARENT")
       
      # bind everything
      tree_info <- rbind(tree_info, children)
      
      # check if there are any open splits left
      do_splits <- !all(tree_info$TERMINAL != "SPLIT")
    } # end for
  } # end while
  
  # calculate fitted values
  leafs <- tree_info[tree_info$TERMINAL == "LEAF", ]
  fitted <- c()
  for (i in seq_len(nrow(leafs))) {
    # extract index
    ind <- as.numeric(rownames(subset(data, eval(parse(text = leafs[i, "FILTER"])))))
    # estimator is the mean y value of the leaf
    fitted[ind] <- mean(y[ind])
  }
  
  # return everything
  return(list(tree = tree_info, fit = fitted, formula = formula, data = data))
}


R/reg_tree_imp.R
# Regression Tree ---------------------------------------------------------

# This is the splitting criterion we minimize (SSE [Sum Of Squared Errors]):
# $SSE = \sum_{i \in S_1} (y_i - \bar(y)1)^2 + \sum_{i \in S_2} (y_i - \bar(y)2)^2$
sse_var <- function(x, y) {
  splits <- sort(unique(x))
  sse <- c()
  for (i in seq_along(splits)) {
    sp <- splits[i]
    sse[i] <- sum((y[x < sp] - mean(y[x < sp]))^2) + sum((y[x >= sp] - mean(y[x >= sp]))^2) 
  }
  split_at <- splits[which.min(sse)]
  return(c(sse = min(sse), split = split_at))
}



#' reg_tree
#' Fits a simple regression tree with SSE splitting criterion. The estimator function
#' is the mean.
#' 
#' @param formula an object of class formula
#' @param data a data.frame or matrix
#' @param minsize a numeric value indicating the minimum size of observations
#'                in a leaf
#'
#' @return \itemize{
#' \item tree - the tree object containing all splitting rules and observations
#' \item imp - returns the feature importance
#' \item fit - our fitted values, i.e. X %*% theta
#' \item formula - the underlying formula
#' \item data - the underlying data
#' }
#' @export
#'
#' @examples # Complete runthrough see: www.github.com/andrebleier/cheapml
#'# Regression Tree Runthrough ----------------------------------------------
#'
#'# load data simulation tool
#'# library(devtools)
#'# devtools::install_github('andrebleier/Xy')
#'library(Xy)
#'
#'
#'# simulate data
#'sim <- Xy(n = 2500, # 2500 observations
#'          numvars = c(3, 0), # 2 linear variables / 0 nonlinear
#'          noisevars = 0, # no noise variables
#'          catvars = 0, # omit dummy variables
#'          stn = 5,
#'          intercept = FALSE
#')
#'
#'# simulation overview
#'sim
#'
#'# get the formula
#'eq <- sim$eq
#'
#'# get the formula
#'model_df <- sim$data
#'
#'
#'# fit
#'mod <- reg_tree_imp(formula = eq, data = model_df, minsize = 100)
#'
#'# our tree
#'mod$tree
#'
#'# the feature importance
#'mod$imp
#'
#'# check the feature importance of the simulation
#'varimp(sim)

reg_tree_imp <- function(formula, data, minsize) {
  
  # coerce to data.frame
  data <- as.data.frame(data)
  
  # handle formula
  formula <- terms.formula(formula)
  
  # get the design matrix
  X <- model.matrix(formula, data)
  
  # extract target
  y <- data[, as.character(formula)[2]]
  
  # initialize while loop
  do_splits <- TRUE
  
  # create output data.frame with splitting rules and observations
  tree_info <- data.frame(NODE = 1, NOBS = nrow(data), FILTER = NA, TERMINAL = "SPLIT",
                          IMP_GINI = NA, SPLIT = NA, stringsAsFactors = FALSE)
  
  # keep splitting until there are only leafs left
  while(do_splits) {
    
    # which parents have to be splitted
    to_calculate <- which(tree_info$TERMINAL == "SPLIT")
    
    for (j in to_calculate) {
      
      # handle root node
      if (!is.na(tree_info[j, "FILTER"])) {
        # subset data according to the filter
        this_data <- subset(data, eval(parse(text = tree_info[j, "FILTER"])))
        # get the design matrix
        X <- model.matrix(formula, this_data)
      } else {
        this_data <- data
      }
      
      # estimate splitting criteria
      splitting <- apply(X,  MARGIN = 2, FUN = sse_var, y = this_data[, all.vars(formula)[1]])
      
      # get the min SSE
      tmp_splitter <- which.min(splitting[1,])
      
      # define maxnode
      mn <- max(tree_info$NODE)
      
      # paste filter rules
      current_filter <- c(paste(names(tmp_splitter), ">=", 
                            splitting[2,tmp_splitter]),
                      paste(names(tmp_splitter), "<", 
                            splitting[2,tmp_splitter]))
      
      # Error handling! check if the splitting rule has already been invoked
      split_here  <- !sapply(current_filter,
                             FUN = function(x,y) any(grepl(x, x = y)),
                             y = tree_info$FILTER)
      
      # append the splitting rules
      if (!is.na(tree_info[j, "FILTER"])) {
        current_filter  <- paste(tree_info[j, "FILTER"], 
                             current_filter, sep = " & ")
      } 
      
      # calculate metrics within the children
      metr <- lapply(current_filter,
                         FUN = function(i, x, data, formula) {
                          df <- subset(x = x, subset = eval(parse(text = i)))
                          nobs <- nrow(df)
                          w <- nobs/nrow(data)
                          y <- df[, all.vars(formula)[1]]
                          imp <- mean((y - mean(y, na.rm = TRUE))^2)
                          return(c(nobs, w*imp))
                         },
                         x = this_data, data = data, formula = formula)  
      
      # extract relevant information
      current_nobs <- sapply(metr, function(x) x[[1]])
      imp_sum_child <- sum(sapply(metr, function(x) x[[2]]))
      current_y <- this_data[, all.vars(formula)[1]]
      imp_parent <- nrow(this_data)/nrow(data) * mean((current_y-mean(current_y))^2)
      imp_gini <- imp_parent - imp_sum_child
      
      # insufficient minsize for split
      if (any(current_nobs <= minsize)) {
        split_here <- rep(FALSE, 2)
      }
      
      # create children data frame
      children <- data.frame(NODE = c(mn+1, mn+2),
                             NOBS = current_nobs,
                             FILTER = current_filter,
                             TERMINAL = rep("SPLIT", 2),
                             IMP_GINI = NA,
                             SPLIT = NA,
                             row.names = NULL)[split_here,]
      
      # overwrite state of current node, add gini importance and split variable
      tree_info[j, "TERMINAL"] <- ifelse(all(!split_here), "LEAF", "PARENT")
      tree_info[j, "IMP_GINI"] <- imp_gini
      if (tree_info[j, "TERMINAL"] == "PARENT") {
      tree_info[j, "SPLIT"] <- names(tmp_splitter)
      }
      
      # bind everything
      tree_info <- rbind(tree_info, children)
      
      # check if there are any open splits left
      do_splits <- !all(tree_info$TERMINAL != "SPLIT")
    } # end for
  } # end while
  
  # calculate fitted values
  leafs <- tree_info[tree_info$TERMINAL == "LEAF", ]
  fitted <- c()
  for (i in seq_len(nrow(leafs))) {
    # extract index
    ind <- as.numeric(rownames(subset(data, eval(parse(text = leafs[i, "FILTER"])))))
    # estimator is the mean y value of the leaf
    fitted[ind] <- mean(y[ind])
  }
  
  # calculate feature importance
  imp <- tree_info[, c("SPLIT", "IMP_GINI")]
  
  if (!all(is.na(imp$SPLIT))) {
  imp <- aggregate(IMP_GINI ~ SPLIT, FUN = function(x, all) sum(x, na.rm = T)/sum(all, na.rm = T), 
                    data = imp, all = imp$IMP_GINI)
  }
  
  # rename to importance
  names(imp) <- c("FEATURES", "IMPORTANCE")
  imp <- imp[order(imp$IMPORTANCE, decreasing = TRUE),]
  
  # return everything
  return(list(tree = tree_info, fit = fitted, formula = formula, 
              importance = imp, data = data))
}


R/gradient_boosting.R
# Gradient Boosting -------------------------------------------------------

# this is our loss function
# y - the target
# yhat - the fitted values
loss <- function(y, yhat) return(mean(1/2*(y-yhat)^2))

# the derivative of our loss function is the gradient
# y - the target
# yhat - the fitted values
gradient <- function(y, yhat) {return(y - yhat)}

# grad_boost
#' Fit a boosted linear model
#'
#' @param formula an object of class formula
#' @param data a data.frame or matrix
#' @param nu a numeric within the range of [0,1], the learning rate
#' @param stop a numeric value determining the total boosting iterations
#' @param loss.fun a convex loss function which is continuously differentiable
#' @param grad.fun a function which computes the gradient of the loss function
#' @param yhat.init a numeric value determining the starting point
#'
#' @return \itemize{
#' \item theta - this is our estimator
#' \item u - this is the last gradient value
#' \item fit - our fitted values, i.e. X %*% theta
#' \item formula - the underlying formula
#' \item data - the underlying data
#' }
#' @export
#'
#' @examples # Complete runthrough see: www.github.com/andrebleier/cheapml
#'# Gradient Boosting Runthrough --------------------------------------------
#'
#'# load data simulation tool
#'# library(devtools)
#'# devtools::install_github('andrebleier/Xy')
#'library(Xy)
#'
#'# simulate data
#'sim <- Xy(n = 1000, # 1000 observations
#'         numvars = c(2, 0), # 2 linear variables / 0 nonlinear
#'         noisevars = 0, # no noise variables
#'         catvars = 0, # omit dummy variables
#'         stn = 30, # signal to noise ratio 30:1
#'         intercept = TRUE
#'         )
#'
#'# simulation overview
#'sim
#'
#'# get the formula
#'eq <- sim$eq
#'
#'# get the formula
#'model_df <- sim$data
#'
#'# source the function
#'source("algorithms/grad_boost.R")
#'
#'# fit
#'mod <- grad_boost(formula = eq, data = model_df,
#'                  nu = 0.1, stop = 100,
#'                  grad.fun = gradient,
#'                  loss.fun = loss)
#'
#'# our estimator
#'mod$theta
#'
#'# the true underlying effect
#'coef(sim)
#'
grad_boost <- function(formula, data, nu = 0.01, stop, 
                       grad.fun, loss.fun, yhat.init = 0) {
  
  # coerce to data.frame
  data <- as.data.frame(data)
  
  # handle formula
  formula <- terms.formula(formula)
  
  # get the design matrix
  X <- model.matrix(formula, data)
    
  # extract target
  y <- data[, as.character(formula)[2]]

  # initial fit
  fit <- yhat.init
  
  # initialize the gradient with yhat.init
  u <- grad.fun(y = y, yhat = fit)
  
  # initial working parameter (our theta)
  # this is just an empty body
  theta <- rep(0, ncol(X))
  
  # track loss
  loss <- c()
  
  # boost from 1 to stop
  for (i in 1:stop) {
    
    # estimate working parameter (via OLS)
    # theta_i = (X'X)^-1 X'y
    # This is the (base procedure) where you can literally place 
    # any optimization algorithm
    base_prod <- lm.fit(x = X, y = u)
    theta_i <- coef(base_prod)
  
    # update our theta
    theta <- theta + nu*as.vector(theta_i)
    
    # new fitted values
    fit <- fit + nu * fitted(base_prod)
    
    # update gradient
    u <- grad.fun(y = y, yhat = fit)
    
    # update loss 
    loss <- append(loss, loss.fun(y = y, yhat = fit))
  }  
  
  # naming the estimator
  names(theta) <- colnames(X)
  
  # define return
  return(list(theta = theta, u = u, fit = fit, loss = loss, 
              formula = formula, data = data))
}



R/reg_rf.R
# Regression Random Forest ------------------------------------------------

#' reg_rf
#' Fits a random forest with a continuous scaled features and target 
#' variable (regression)
#'
#' @param formula an object of class formula
#' @param n_trees an integer specifying the number of trees to sprout
#' @param feature_frac an numeric value defined between [0,1]
#'                     specifies the percentage of total features to be used in
#'                     each regression tree
#' @param data a data.frame or matrix
#'
#' @importFrom plyr raply
#' @return
#' @export
#'
#' @examples # Complete runthrough see: www.github.com/andrebleier/cheapml
#'
#' 
#' load data simulation tool
#' library(devtools)
#' devtools::install_github('andrebleier/Xy')
#'library(Xy)
#'library(ggplot2)
#'source("algorithms/reg_rf.R")
#'source("algorithms/reg_tree_imp.R")
#'
#'# simulate data
#'sim <- Xy(n = 2500, # 2500 observations
#'          numvars = c(5, 0), # 5 linear variables / 0 nonlinear
#'          noisevars = 5, # 5 noise variables
#'          catvars = 0, # omit dummy variables
#'          weights = c(-5, 20), # range of the beta weights
#'          stn = 5, # signal to noise ratio
#'          intercept = FALSE # intercept in the model
#')
#'
#'# simulation overview
#'sim
#'
#'# get the formula
#'eq <- sim$eq
#'
#'# get the formula
#'model_df <- sim$data
#'
#'# plot the true variable importance
#'imp_true <- varimp(sim, plot = FALSE)
#'
#'# subset to importance mean
#'imp_true <- imp_true[, c(1,2)]
#'names(imp_true)[2] <- "IMPORTANCE" 
#'imp_true$TYPE <- "Truth"
#'
#'# fit regression tree
#'mod_rt <- reg_tree_imp(formula = eq, data = model_df, minsize = 250)
#'
#'# fit random forest
#'mod_rf <- reg_rf(formula = eq,
#'                 n_trees = 50, feature_frac = 0.8, data = model_df)
#'
#'# importance regression tree
#'imp_rt <- mod_rt$importance
#'imp_rt$TYPE <- "Regression Tree"
#'  
#'# importance random forest
#'imp_rf <- mod_rf$importance
#'imp_rf$TYPE <- "Random Forest"
#'
#'# merge to a plot_df
#'plot_df <- rbind(imp_rf, imp_rt, imp_true)
#'
#'ggplot(plot_df, aes(x = FEATURES, y = IMPORTANCE, fill = TYPE)) + 
#'  geom_bar(stat = "identity", position = "dodge") +
#'  labs(x = "") +
#'  theme_minimal() +
#'  coord_flip() +
#'  scale_fill_manual(name = "", values = c("#013848", "#00A378", "#FF8000"))
  
reg_rf <- function(formula, n_trees, feature_frac, data) {
  
  
  # load plyr
  require(plyr)
  
  # define function to sprout a single tree
  sprout_tree <- function(formula, feature_frac, data) {
    # extract features
    features <- all.vars(formula)[-1]
    
    # extract target
    target <- all.vars(formula)[1]
    
    # bag the data
    # - randomly sample the data with replacement (duplicate are possible)
    temp1 <- sample(1:nrow(data), size = nrow(data), replace = TRUE)
    train <- data[temp1,]

    # randomly sample features
    # - only fit the regression tree with feature_frac * 100 % of the features
    features_sample <- sample(features,
                              size = ceiling(length(features) * feature_frac),
                              replace = FALSE)
    
    # create new formula
    formula_new <-
      as.formula(paste0(target, " ~ -1 + ", paste0(features_sample,
                                              collapse =  " + ")))
    
    # fit the regression tree
    tree <- reg_tree_imp(formula = formula_new,
                         data = train,
                         minsize = ceiling(nrow(train) * 0.1))

    #the fitted values in tree$fit is built by averaging  over leafs. It can have NA where values were missing from the boostrapped sample.
    # If these msiing values were at the end of the data set then the returned tree$fit will have the wrong length
    if( length(tree$fit) != dim(train)[[1]]){
        t1<- dim(train)[[1]] -length(tree$fit)
        tree$fit <-c(tree$fit,rep(NA,t1))
    }
    
    aa<-data.frame(tree$fit,temp1) %>%
        dplyr::group_by(temp1)%>%
        dplyr::summarize(Mean = mean(tree.fit, na.rm=TRUE),.groups = 'drop')
    aa<-as.matrix(aa)
    tt<-rep(NA,length(temp1))
    tt[aa[,1]]<-aa[,2]
    tree$fit <- tt
      
    # save the fit and the importance
    return(list(tree$fit, tree$importance))
  }
  
  # apply the rf_tree function n_trees times with plyr::raply
  # - track the progress with a progress bar
  trees <- plyr::raply(
    n_trees,
    sprout_tree(
      formula = formula,
      feature_frac = feature_frac,
      data = data
    ),
    .progress = "text"
  )
  
  # extract fit
  fits <- do.call("cbind", trees[, 1])
  
  # calculate the final fit as a mean of all regression trees
  rf_fit <- apply(fits, MARGIN = 1, mean, na.rm = TRUE)

  # extract the feature importance
  imp_full <- do.call("rbind", trees[, 2])
  
  # build the mean feature importance between all trees
  imp <- aggregate(IMPORTANCE ~ FEATURES, FUN = mean, imp_full)
  
  # build the ratio for interpretation purposes
  imp$IMPORTANCE <- imp$IMPORTANCE / sum(imp$IMPORTANCE)
  
  # export
  return(list(fit = rf_fit,
              importance = imp[order(imp$IMPORTANCE, decreasing = TRUE), ]))
}
